engine/src/main/java/org/pentaho/di/trans/steps/ldapinput/LDAPConnection.java

```diff
@@ -0,0 +1,771 @@
+                new PagedResultsControl( GetPagingSize(), cookie, Control.CRITICAL ) } );
+          } else {
+            getInitialContext().setRequestControls(
+              new Control[] { new PagedResultsControl( GetPagingSize(), cookie, Control.CRITICAL ) } );
+          }
+          if ( ( cookie != null ) && ( cookie.length != 0 ) ) {
+            // get search result for the page
+            this.results = getInitialContext().search( getSearchBase(), getFilter(), getSearchControls() );
+          } else {
+            return null;
+          }
+
+        } catch ( Exception e ) {
+          throw new KettleException( BaseMessages.getString( PKG, "LDAPInput.Exception.ErrorPaging" ), e );
+        }
+
+        while ( !getSearchResult().hasMoreElements() ) {
+          return null;
+        }
+      } else {
+        // User do not want to use paging
+        // we have already returned all the result
+        return null;
+      }
+    }
+
+    try {
+      SearchResult searchResult = getSearchResult().next();
+      Attributes results = searchResult.getAttributes();
+      results.put( "dn", searchResult.getNameInNamespace() );
+      return results;
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG, "LDAPConnection.Exception.GettingAttributes" ), e );
+    }
+  }
+
+  private InitialLdapContext getInitialContext() {
+    return protocol.getCtx();
+  }
+
+  private SearchControls getSearchControls() {
+    return this.controls;
+  }
+
+  private NamingEnumeration<SearchResult> getSearchResult() {
+    return this.results;
+  }
+
+  /**
+   * Remove CR and LF from filter string
+   *
+   * @param filter
+   * @return corrected filter
+   */
+  private static String correctFilter( String filter ) {
+    return Utils.isEmpty( filter ) ? "" : filter.replaceAll( "(\\r|\\n)", "" );
+  }
+
+  public static String extractBytesAndConvertToString( Attribute attr, boolean isSID ) throws Exception {
+    byte[] b;
+    try {
+      b = (byte[]) attr.get();
+    } catch ( Exception e ) {
+      // Get bytes from String
+      b = attr.get().toString().getBytes();
+    }
+    if ( isSID ) {
+      return getSIDAsString( b );
+    } else {
+      return byteToHexEncode( b );
+    }
+  }
+
+  /**
+   * Convert the SID into string format
+   *
+   * @param SID
+   * @return String representation of SID
+   */
+  private static String getSIDAsString( byte[] SID ) {
+    long version;
+    long authority;
+    long count;
+    long rid;
+    String strSID;
+    strSID = "S";
+    version = SID[0];
+    strSID = strSID + "-" + Long.toString( version );
+    authority = SID[4];
+    for ( int i = 0; i < 4; i++ ) {
+      authority <<= 8;
+      authority += SID[4 + i] & 0xFF;
+    }
+    strSID = strSID + "-" + Long.toString( authority );
+    count = SID[2];
+    count <<= 8;
+    count += SID[1] & 0xFF;
+    for ( int j = 0; j < count; j++ ) {
+      rid = SID[11 + ( j * 4 )] & 0xFF;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).